### PR TITLE
[ENSVAR-4348]- stop_retained

### PR DESCRIPTION
**Pull request by olaaustine**
[ENSVAR-4348](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-4348)

### CODE REVIEW

File: modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm [hunk 1]
```diff
@@ -1390,6 +1417,7 @@ sub frameshift {
     if($bvfoa->isa('Bio::EnsEMBL::Variation::TranscriptVariationAllele')) {
 
         return 0 if partial_codon(@_);
+        return 0 if stop_retained(@_);
```
Comment (nuno-agostinho) [inline modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm:1420]: Can't it be both `frameshift` and `stop_retained`? I couldn't come up with any examples for this, so maybe it's fine.

**Review by nuno-agostinho — CHANGES_REQUESTED** [review]
Hi @olaaustine, thanks for the PR.

Can you please add unit tests with some of the examples you have in [ENSVAR-4348](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-4348)?

Here are some cases that I think that we should discuss (@jamie-m-a, @sarahhunt):

| Variant                | Transcript | Comment |Codons | Amino acids | PR result | Expected (in my opinion)   | 
|-----------------|---|--|-----------|----|------|----------------------------|
| `6 30590698 . GTGA CTAG` || Missense mutation + synonymous stop codon | `gaGTGA/gaCTAG` | `E*/D*` | `missense` | `missense,stop_retained` |
| `5 134750948 . G -` || Deletion of one base in the stop codon, although still retained because of the next base | `tGa/ta` | `*/X` | `frameshift_variant,stop_lost` | `stop_retained` |
| `6 30590700 . G -` || Deletion of one base in the stop codon, although still retained because of the next base | `tGa/ta` | `*/X` | `frameshift_variant,stop_lost` | `stop_retained` |
| `6 30590698 . G GTAG` || Insertion of a stop codon before the stop codon | `-/TAG` | `-/*` | `stop_gained,inframe_insertion` | `inframe_insertion,stop_retained` |
| `1 225419264 . A AG` | ENST00000425080 | Insertion near an incomplete terminal codon | `cct/ccCt` | `P/PX` | `inframe_insertion,stop_retained_variant` | `incomplete_terminal_codon_variant` |
| `1 160355719 . C CCTGAAGAGGACCTGAACTTTGTAA` | ENST00000424754 | Insertion near an incomplete terminal codon | `cct/cCTGAAGAGGACCTGAACTTTGTAAct` | `P/PEEDLNFVT` | `inframe_insertion,stop_retained_variant` | `inframe_insertion,incomplete_terminal_codon_variant ` |
| `1 161170462 . G GT` | ENST00000539753 | Insertion causes frameshift and stop loss | `gta/gTta` | `V/VX` | `inframe_insertion,stop_retained_variant,NMD_transcript_variant` | `frameshift_variant,stop_lost,NMD_transcript_variant` |
| `1 156135949 1768448 C CG` | ENST00000675881 | Insertion causes frameshift and stop loss | `ccg/ccGg` | `P/PX` | `inframe_insertion,stop_retained_variant,NMD_transcript_variant` | frameshift_variant,stop_lost,NMD_transcript_variant |

**Review by nuno-agostinho — APPROVED** [review]
Most edge cases have been fixed, thanks @olaaustine!

File: modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm [hunk 2]
```diff
@@ -1465,11 +1465,14 @@ sub partial_codon {
         $cache->{_partial_codon} = 0;
 
         $bvfo ||= $bvfoa->base_variation_feature_overlap;
+
+        my ($ref_pep, $alt_pep) = _get_peptide_alleles(@_);
```
[the remark this answers is not in the document]
Comment (olaaustine) [reply modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm]: Hi @nuno-agostinho, 
That line has been removed. Thank you.
Ola

**Comment by nuno-agostinho** [comment]
Merged to `main` and `release/112` branches.

**Pull request merged**